craft_parts/sequencer.py

```diff
@@ -181,6 +191,29 @@ def _run_step(
     ) -> None:
         self._process_dependencies(part, step)
 
+        if step == Step.OVERLAY:
+            # Make sure all previous layers are in place before we add a new
+            # layer to the overlay stack,
+            layer_hash = self._ensure_overlay_consistency(
+                part,
+                reason=f"required to overlay {part.name!r}",
+                skip_last=True,
+            )
+            self._layer_state.set_layer_hash(part, layer_hash)
+
+        elif (step == Step.BUILD and part in self._overlay_viewers) or (
+            step == Step.STAGE and part.has_overlay
+        ):
+            # The overlay step for all parts should run before we build a part
+            # with overlay visibility or before we stage a part that declares
+            # overlay parameters.
+            last_part = self._part_list[-1]
+            verb = _step_verb[step]
+            self._ensure_overlay_consistency(
+                last_part,
+                reason=f"required to {verb} {part.name!r}",
+            )
```

Comment:
Later, in the build or stage steps, we verify if the layer stack is consistent based on the verification hash.